### PR TITLE
throw DataNotFound error with clear msg

### PR DESCRIPTION
**Pull request by RichardLi1437**

### CODE REVIEW

File: assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py [hunk 1]
```diff
@@ -126,7 +126,8 @@ def _raw_mdc_uri_folder_to_mltable(
     return table
 
 
-def _convert_mltable_to_spark_df(table: MLTable, preprocessed_input_data: str, fs: AbstractFileSystem) -> DataFrame:
+def _convert_mltable_to_spark_df(table: MLTable, preprocessed_input_data: str,
+                                 fs: AbstractFileSystem = None, add_tags_func=None) -> DataFrame:
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py:130]: Why do we need to pass in the function?

File: assets/model_monitoring/components/src/shared_utilities/event_utils.py [hunk 2]
```diff
@@ -69,17 +69,31 @@ def _get_timestamp():
...
+def add_tags_to_run(run_id: str, tags: dict):
+    """Add tags to the run with the specified run id."""
+    client = mlflow.tracking.MlflowClient()
+    for key, value in tags.items():
+        client.set_tag(run_id, key, value)
+
+
+def add_tags_to_root_run(tags: dict):
+    """Add tags to the root run."""
+    run = mlflow.get_run(run_id=_get_run_id())
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/shared_utilities/event_utils.py:81]: do you have this function, did you test it?

File: assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py [hunk 3]
```diff
@@ -146,7 +147,13 @@ def _convert_mltable_to_spark_df(table: MLTable, preprocessed_input_data: str, f
         )
 
     # Read mltable from preprocessed_data
-    return try_read_mltable_in_spark(des_path, "preprocessed_data")
+    try:
+        return try_read_mltable_in_spark_with_error(des_path, "preprocessed_data")
+    except DataNotFoundError as e:
+        tags = {"no_input_data": True}
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py:153]: Can we follow the current tag prefix such as "AML_SEND_EMAIL_ADDRESSES_TAG"

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py [hunk 4]
```diff
@@ -14,45 +24,60 @@ def init_spark():
...
-    return (
-        isinstance(error, IndexError)
-        or "The requested stream was not found" in error.args[0]
-        or "Not able to find MLTable file" in error.args[0]
-    )
+def _is_not_found_exception(input_name, error: Exception):
+    if isinstance(error, IndexError) or ("Partition 0 is out of bounds." in error.args[0]):
+        return True, (f"No data is found for input '{input_name}' in the specified window, "
+                      "most likely there is no request in the specified window.")
+    elif "The requested stream was not found" in error.args[0]:
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/shared_utilities/io_utils.py:31]: Do we know which data is missing?

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py [hunk 5]
```diff
@@ -14,45 +24,60 @@ def init_spark():
...
+            raise DataNotFoundError(error_message)
 
-def try_read_mltable_in_spark(mltable_path: str, input_name: str) -> DataFrame:
-    """Read mltable in spark. In case of failure, returns None."""
     try:
-        return read_mltable_in_spark(mltable_path)
+        df = read_mltable_in_spark(mltable_path)
     except Exception as error:
-        if _is_not_found_exception(error):
+        input_not_found, error_message = _is_not_found_exception(input_name, error)
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/shared_utilities/io_utils.py:71]: Should we put mltable path in the error message to let the customer debug themselves?

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py [hunk 6]
```diff
@@ -14,45 +24,60 @@ def init_spark():
...
+        if input_not_found:
             print(error)
-            error_message = f"No data was found for input '{input_name}'."
-            print(error_message)
+            return process_input_not_found(error_message)
         else:
             raise error
+    if (not df) or df.isEmpty():
+        error_message = (f"No data is found for input '{input_name}' in the specified window, "
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/shared_utilities/io_utils.py:78]: can we also print out the time window string

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py [hunk 7]
```diff
@@ -14,45 +24,60 @@ def init_spark():
...
-        return read_mltable_in_spark(mltable_path)
+        df = read_mltable_in_spark(mltable_path)
     except Exception as error:
-        if _is_not_found_exception(error):
+        input_not_found, error_message = _is_not_found_exception(input_name, error)
+        if input_not_found:
             print(error)
-            error_message = f"No data was found for input '{input_name}'."
-            print(error_message)
+            return process_input_not_found(error_message)
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/shared_utilities/io_utils.py:74]: do we need both "process_input_not_found" and "_is_not_found_exception"?

Comment (yunjie-hub) [reply assets/model_monitoring/components/src/shared_utilities/io_utils.py:74]: We just need one for error handling right?

File: assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py (hunk 1 above)
Comment (RichardLi1437) [reply assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py:130]: follow the param injection patten to inject a mock add_tags_func for testing

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py (hunk 7 above)
Comment (RichardLi1437) [reply assets/model_monitoring/components/src/shared_utilities/io_utils.py:74]: process_input_not_found()  is to handle the input not found excepton, while _is_not_found_exception() is to check if the exception indicates an input not found error.

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py (hunk 4 above)
Comment (RichardLi1437) [reply assets/model_monitoring/components/src/shared_utilities/io_utils.py:31]: added mltable path to the error msg so user can check the MLTable content to get all the details.

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py (hunk 5 above)
Comment (RichardLi1437) [reply assets/model_monitoring/components/src/shared_utilities/io_utils.py:71]: added

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py (hunk 6 above)
Comment (RichardLi1437) [reply assets/model_monitoring/components/src/shared_utilities/io_utils.py:78]: time window string is applied in _raw_mdc_uri_folder_to_mltable(), after that, only mltable is passed down, all the detail including time window is in the MLTable. Added mltable path to the error msg, user can check it by themselves.

File: assets/model_monitoring/components/model_monitor/model_data_collector_preprocessor/spec.yaml [hunk 8]
```diff
@@ -4,7 +4,7 @@ type: spark
 name: model_data_collector_preprocessor
 display_name: Model Data Collector - Preprocessor
 description: Filters the data based on the window provided.
-version: 0.3.9
+version: 0.3.10
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/model_monitor/model_data_collector_preprocessor/spec.yaml:7]: Do you need to upgrade other components version?

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py [hunk 9]
```diff
@@ -14,45 +34,76 @@ def init_spark():
...
+    Read mltable in spark.
+
+    If data not found, conduct different error handling based on no_data_approach.
+    """
+    def process_input_not_found(input_not_found_category: InputNotFoundCategory):
+        err_msg_map = {
+            InputNotFoundCategory.NOT_INPUT_MISSING:
+                "Not input missing error.",
+            InputNotFoundCategory.NO_INPUT_IN_WINDOW: (
+                f"No data is found for input '{input_name}' in the specified window, "
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/shared_utilities/io_utils.py:69]: Can we have the error message that what directory we are looking for, so customer should know it's the input data issue

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py [hunk 10]
```diff
@@ -14,45 +34,76 @@ def init_spark():
...
+                "most likely there is no request in the specified window. "
+                f"You can check the filter field of MLTable in {mltable_path} for the time window."
+            ),
+            InputNotFoundCategory.ROOT_FOLDER_NOT_FOUND: (
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/shared_utilities/io_utils.py:73]: This is the same, is it possible to send back the path that we are trying to read

File: assets/model_monitoring/components/model_monitor/model_data_collector_preprocessor/spec.yaml (hunk 8 above)
Comment (RichardLi1437) [reply assets/model_monitoring/components/model_monitor/model_data_collector_preprocessor/spec.yaml:7]: yes, but I'd like to do it in a separate PR and ask all component owners to sign off

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py (hunk 9 above)
Comment (RichardLi1437) [reply assets/model_monitoring/components/src/shared_utilities/io_utils.py:69]: all the input detail is in mltable, we don't have it in this context, but I listed the MLTable path in the err msg, user can check the MLTable contents for all the detail. For this particular no data in window, I also highlighted the filter field for user to check the time window.

File: assets/model_monitoring/components/src/shared_utilities/io_utils.py (hunk 10 above)
Comment (RichardLi1437) [reply assets/model_monitoring/components/src/shared_utilities/io_utils.py:73]: again, the path info is in spark mltable utilities which is out of MoMo code base, and I've included mltable_path in the err msg and highlighted pattern field for the customer to check.

**Review by yunjie-hub — APPROVED** [review]

**Pull request merged**